Project 9

VariusValinium22 opened the pull request:
First Pull request asking to merge these changes into the main branch

VariusValinium22 committed:
deploy a new project-9 Github Pages branch

VariusValinium22 committed:
moved index.html out of the src folder so the github Pages won't firstly show README.md

VariusValinium22 committed:
moved index.html back in since it didn't work

VariusValinium22 committed:
created getUserInfo and getInitialCards functions

VariusValinium22 committed:
changed api.js

VariusValinium22 committed:
changed api.js again

VariusValinium22 committed:
adding avatar module

VariusValinium22 committed:
got the avatar picture to update

VariusValinium22 committed:
can't get the userInfo to stay on the page after changing it in the edit modal

VariusValinium22 committed:
fixed the error by adding profileTitle and profileDescripion imports to index.js

VariusValinium22 committed:
finished setting up POST fetch for addNewCard() and the newCardPopup

VariusValinium22 committed:
Pro9: finished all the tasks

VariusValinium22 committed:
Pro9: pushing for review

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,100 @@
+  getUserInfo() {
+    return fetch(`${this.baseUrl}/users/me`, {
+      headers: this.headers,
+    }).then(this._handleRequest);
```
**COULD BE IMPROVED**

You can make a universal method `_request` not to duplicate the `ok` checking, `headers` and `baseUrl`


```js
 _request(endpoint, options = {}) {
    const finalOptions = {
      // add the headers
      headers: this._headers,
      // spread the other options if provided
      ...options,
    };
      // add the endpoint to the base url
    const url = `${this._baseUrl}${endpoint}`;
    return fetch(url, finalOptions).then(this._checkResponse);
  }

  // usage

  getUserInfo() {
    // pass only the endpoint here
    return this._request("/users/me");
  }

  getInitialCards() {
    // pass only the endpoint here
    return this._request("/cards");
  }

  // and so on



```

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,100 @@
+      method: "PATCH",
+      headers: {
+        ...this.headers,
+        "Content-Type": "application/json",
```
According to the task you need to pass full `headers` into the constructor of `Api`. See here https://snipboard.io/pRdaAZ.jpg . And in the code you need to use only: `headers: this._headers` (not to duplicate the code)

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,100 @@
+      },
+      body: JSON.stringify(cardData),
+    })
+      .then((response) => response.json())
+      .then(this._handleRequests);
+  }
```
You have `.then(this._handleRequest);` for the `ok` checking. Please, use it everywhere not to duplicate the code


You don't have `_handleRequests` in your code

VariusValinium22 replied:
I have _handleRequests() method at the end of all fetch request methods.

  _handleRequest(res) {
    return res.ok ? res.json() : Promise.reject(`Error: ${res.status}`);
  }

gennady-bars replied:
> I have _handleRequests() method at the end of all fetch request methods.
> 
> _handleRequest(res) { return res.ok ? res.json() : Promise.reject(`Error: ${res.status}`); }


Pay attantion to the `s` at the end of the method `_handleRequests`. You don't have it in your code. You have `_handleRequest` without `s`

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,100 @@
+
+export { api };
+
+api.getUserInfo();
```
`api.getUserInfo();` is not needed here. Everything is done in `index.js`

gennady-bars commented on src/components/Card.js:
```diff
@@ -1,24 +1,29 @@
   _setEventListeners() {
     //card__like-button
     this._cardElement
       .querySelector(".card__like-button")
```
**Could be improved**

All elements found **every time** you call a method should be defined as **class fields** in the constructor, `_setEventListeners` (or in `getView`) only **1 time** in order not to waste resources on searching them again and again.

`Class fields` (the so-called **class variables** with `this`) are accessible inside any class method. You shouldn’t pass **class fields** into the call of any method or search them again because you can use class fields inside all methods.
There can be `this._inputs`, `this._submitButton`, `this._cardImage`, `this._form`, `this._likeButton`, and so on.

```js
  getView() {
    this._element = this._getTemplate();
    this._likeButton = this._element.querySelector(".card__like-button");
    this._cardImage = this._element.querySelector(".card__image");
    this._trashButton = this._element.querySelector(".card__trash-button");

```

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,24 +1,29 @@
+import PopupWithDelete from "../components/PopupWithDelete.js";
 import Section from "../components/Section.js";
 import Card from "../components/Card.js";
 import FormValidator from "../components/FormValidator.js";
```
Please, delete `this._toggleButtonState();` from line 23 in `FormValidator.js`

You need to **disable** the button only in a `then` block after a **successful response** to be able to click it again if there is a server error.

gennady-bars commented on src/components/PopupWithDelete.js:
```diff
@@ -0,0 +1,32 @@
+    this._buttonText = this._submitButton.textContent;
+  }
+
+  setSubmitAction(action) {
+    this._handleFormSubmit = action;
+  }
```
Nice `setSubmitAction` method

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -17,6 +19,14 @@ class PopupWithForm extends Popup {
     return formValues;
   }
 
+  setButtonText(isLoading, loadingText = "Saving...") {
+    if (isLoading) {
+      this._submitButton.textContent = loadingText;
+    } else {
+      this._submitButton.textContent = this._buttonText;
```
Nice `setButtonText` method

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -17,6 +19,14 @@ class PopupWithForm extends Popup {
+
   setEventListeners() {
     super.setEventListeners();
 
```
```js
      this._popupForm.reset();
      this.close();
```

Please, delete these 2 lines of code from lines 36-37


You need to **reset** the form only   in a `then` block in order not to do that immediately after `submit` because when there is an error, the modal will still be opened and the user can try to send the data again to the server without having to type the data into the inputs once more.

------------------

You need to close modals only in `then` blocks after a **successful response** from the server so that when there is an error, it would be opened (the inputs will not be cleared and the user could send the data again to the server), and if you change the button text, you could see the change as well. **Please, correct it everywhere in the code**

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,24 +1,29 @@
 import FormValidator from "../components/FormValidator.js";
 import "../pages/index.css";
 
+const editAvatarValidator = new FormValidator(config, avatarFormElement);
```
**Could be improved**

You can **universally** create instances of validators for all forms in the project storing them inside one object: `formValidators`.  Then you can take any validator using attribute `name` of the form where you need to disable/enable the submit button or remove errors.

```js
// define an object for storing validators
const formValidators = {}

const enableValidation = (config) => {
  const formList = Array.from(document.querySelectorAll(config.formSelector))
  formList.forEach((formElement) => {
    const validator = new FormValidator(config, formElement)
    // Here you get the name of the form (if you don’t have it then you need to add it into each form in `index.html` first)
    const formName = formElement.getAttribute('name')

   // Here you store the validator using the `name` of the form
    formValidators[formName] = validator;
    validator.enableValidation();
  });
};

enableValidation(config);

```

And now you can use them for disabling buttons or clearing errors:

```js

formValidators[ profileForm.getAttribute('name') ].resetValidation()

// or you can use a string – the name of the form (you know it from `index.html`)

formValidators['profile-form'].resetValidation()

```

gennady-bars commented on src/pages/index.js:
```diff
@@ -30,34 +35,57 @@ addCardValidator.enableValidation();
+      console.error("Error adding new card: ", error);
+    })
+    .finally(() => newCardPopup.setButtonText());
   addCardValidator.disableSubmitButton();
```
`  addCardValidator.disableSubmitButton();
`

* You need to **disable** the button only in a `then` block after a **successful response** to be able to click it again if there is a server error.

gennady-bars commented on src/pages/index.js:
```diff
@@ -67,36 +95,151 @@ addNewCardButton.addEventListener("click", () => {
+const profilePopup = new PopupWithForm("#edit-profile-modal", (formData) => {
   const name = formData["name"];
   const job = formData["description"];
+  profilePopup.setButtonText(true);
```
**COULD BE IMPROVED**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```

gennady-bars commented on src/pages/index.js:
```diff
@@ -67,36 +95,151 @@ addNewCardButton.addEventListener("click", () => {
+  jobInput.value = currentUserInfo.description;
+  editFormValidator.resetValidation();
+
   profilePopup.open();
```
`  profilePopup.open();
`

You have the same code in line 144. Please, delete this one

gennady-bars commented on src/pages/index.js:
```diff
@@ -67,36 +95,151 @@ addNewCardButton.addEventListener("click", () => {
+//==================================
+function handleLikeClick(card) {
+  // if card isLiked removeLike
+  if (card._isLiked) {
```
`_isLiked`

Please, delete the **low dash** (`underscore`) from methods/fields if you want to use them **outside** a class because the `low dash` shows a **private** field which is used only inside the class.

gennady-bars commented on src/pages/index.js:
```diff
@@ -67,36 +95,151 @@ addNewCardButton.addEventListener("click", () => {
+    api
+      .removeLike(card._id)
+      .then(() => {
+        card._handleLikeIcon();
```
Please, delete the **low dash** (`underscore`) from methods/fields if you want to use them **outside** a class because the `low dash` shows a **private** field which is used only inside the class.

gennady-bars commented on src/pages/index.js:
```diff
@@ -67,36 +95,151 @@ addNewCardButton.addEventListener("click", () => {
+      .then(() => {
+        card._handleLikeIcon();
+      })
+      .catch((error) => {
+        console.log("Error removing like: ", error);
```
Good job that you `catch` possible errors at the end of server requests. You can simplify it with `.catch(console.error);` because `console.error` is meant to log errors

gennady-bars reviewed (commented):
**↓Expand to see the whole summary↓**

Hello, Martin!

Great job on submitting your project 9! This time you’ve implemented `Api` integration. So you can create your own simple sites from now on! 

**Parts I really liked in your project:**

* Good job that you `catch` possible errors at the end of server requests.
* Good job that you return the button text in `finally`
* Nice `setButtonText` method

There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.

**Needs correcting:**

* All comments in the code  and the comments below:
* When I reload the page, all **black** hearts become **white** again.
* According to the task you need to pass full `headers` into the constructor of `Api`. See here https://snipboard.io/pRdaAZ.jpg . And in the code you need to use only: `headers: this._headers` (not to duplicate the code)
* You need to **disable** the button only in a `then` block after a **successful response** to be able to click it again if there is a server error.
* You need to **reset** the form only   in a `then` block in order not to do that immediately after `submit` because when there is an error, the modal will still be opened and the user can try to send the data again to the server without having to type the data into the inputs once more.
* You need to close modals only in `then` blocks after a **successful response** from the server so that when there is an error, it would be opened (the inputs will not be cleared and the user could send the data again to the server), and if you change the button text, you could see the change as well. **Please, correct it everywhere in the code**
* Please, delete the **low dash** (`underscore`) from methods/fields if you want to use them **outside** a class because the `low dash` shows a **private** field which is used only inside the class.


There are some comments in your project that can help you make it even better. Please check the `Could be improved` comments.

**A reminder:** It’s necessary to fix all comments to get the project accepted. And `Could be improved` comments are optional, they will be up to you if you want to implement them.

Your hard work on this project is truly commendable. While there are still some adjustments needed. If you need any assistance with the feedback or understanding certain concepts, our tutors are here to help. Keep moving forward and maintain your dedication — you are doing an amazing job!✨

VariusValinium22 commented:
I believe I got all the comments done here. I have a _handleRequest() method at the bottom of the Api.js file that takes care of the res.ok or Promise.reject.

VariusValinium22 committed:
Pro9: commit for review. Fixed all comments except for 1. _handleRequest() is method in Api.js

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,93 @@
+      headers: this.headers,
+      body: JSON.stringify(cardData),
+    })
+      .then((response) => response.json())
+      .then(this._handleRequests);
```
Please, delete these 2 lines and use `.then(this._handleRequest);` here (without `s`)

gennady-bars commented on src/pages/index.js:
```diff
@@ -67,36 +97,152 @@ addNewCardButton.addEventListener("click", () => {
+        name: updatedData.name,
+        about: updatedData.about,
+      });
+      addCardValidator.disableSubmitButton();
```
`addCardValidator.disableSubmitButton()` is for the `card` form. You need to use the `profile validator` here

gennady-bars commented on src/pages/index.js:
```diff
@@ -67,36 +97,152 @@ addNewCardButton.addEventListener("click", () => {
+    .updateAvatar(avatarUrl)
+    .then((avatarData) => {
+      userInfo.setUserAvatar(avatarData);
+      addCardValidator.disableSubmitButton();
```
`addCardValidator.disableSubmitButton()` is for the `card` form. You need to use the `avatar validator` here

gennady-bars reviewed (commented):
Everything is almost fine, but:

* You should **clear the inputs** after a successful adding of a new card to let the user add the 2nd card again without having to remove the old data manually. 
* When I **dislike** a card and reload the page, this card is still **liked** (with a black heart)

VariusValinium22 commented:
Thank you. I believe I've completed these mistakes.

VariusValinium22 committed:
Pro9: changed to the proper validators on index.js. took the s out of .then(this._handleRequests)

gennady-bars reviewed (commented):
(**Please, expand the general comment ↓**)

Dear Martin, 

unfortunately I have to reject the project without review:


* You haven't corrected my remarks from here https://github.com/VariusValinium22/se_project_aroundtheus/pull/1#pullrequestreview-2341794710

VariusValinium22 committed:
issue with the isLiked button remaining true/black after toggling it to false/white and refreshing the page

VariusValinium22 committed:
fixed the new card modal to stop populating the previous field values. Fixed the isLiked button to stop reverting to true/black after refreshing the page

VariusValinium22 commented:
I apologize. I have fixed these two issues. 
-clear the inputs after a successful adding of a new card having removed the old data.
-disliking a card and reloading the page, the cards load properly on the last state of the isLiked value.

gennady-bars commented on src/pages/index.js:
```diff
@@ -30,73 +35,224 @@ addCardValidator.enableValidation();
 
 addNewCardButton.addEventListener("click", () => {
   newCardPopup.open();
+  addCardFormElement.reset();
```
`addCardFormElement.reset();` should be in `then` in line 84

You should **clear ** the inputs only after **submitting** a card in `then` rather than when you open or close popups because if you unexpectedly close the popup, you will be upset if the inputs are empty when you open it again, and you’ll have to type the data again.

You need to **reset** the form only in a `then` block  because when there is an error, the modal will still be opened and the user can try to send the data again to the server without having to type the data into the inputs once more.

gennady-bars reviewed (commented):
Everything is almost fine, but:

* You should **clear ** the inputs only after **submitting** a card in `then` rather than when you open or close popups because if you unexpectedly close the popup, you will be upset if the inputs are empty when you open it again, and you’ll have to type the data again.

VariusValinium22 commented:
I've moved line 98 to line 84 just before closing the newCardPopup modal.

gennady-bars commented:
> I've moved line 98 to line 84 just before closing the newCardPopup modal.

You've forgotten to **push** the changes to github, I suppose

gennady-bars reviewed (commented):
One more step is needed:

* You've forgotten to **push** the changes to github, I suppose. Nothing has been changed in the repository

VariusValinium22 commented:
Ok. I've pushed it again. There should be only a few console.log() functions in the code.

gennady-bars commented:
> Ok. I've pushed it again. There should be only a few console.log() functions in the code

Nothing has changed on github https://snipboard.io/y8Xk5g.jpg

gennady-bars reviewed (commented):
Please, commit and push the changes:


* Nothing has changed on github https://snipboard.io/y8Xk5g.jpg

VariusValinium22 committed:
commiting the change of line 98 to line 84. I've also added these two lines to test deployment  // I just added this text to see if my deployment is working
      console.log(createdCard);

VariusValinium22 commented:
I've added these two lines to the deployment: 
      // I just added this text to see if my deployment is working
      console.log(createdCard);

I've ran these 3 cmds in GitBash:
git add -A
git -m "git commit -m "commiting the change of line 98 to line 84. I've also added these two lines to test deployment  // I just added this text to 
see if my deployment is working
      console.log(createdCard);"
git push      

I see in my window:
VariusValinium22 deployed to github-pages 1 minute ago -- with GitHub Pages.
there is a View deployment button right next to this statement. 
      
      Can you please check that it worked?

gennady-bars commented:
> Can you please check that it worked?

Now I see the commit 
![image](https://github.com/user-attachments/assets/c872aec7-ff33-4de4-9874-a1a22b7e9ea9)

gennady-bars reviewed (approved):
You’ve done a great job!

Your project has been accepted. 

Good luck!

VariusValinium22 committed:
some small changes
